Initial draft changes for converting to Backdrop

yorkshire-pudding opened the pull request:
Hi @alvinjohnsonso 
I'm not quite sure what to do with the menu.
Also, I still don't think I've got the changes in 111-115 right as it doesn't work.

Grateful if you could help.

Thanks
Martin

yorkshire-pudding committed:
Update to meet Backdrop requirements

yorkshire-pudding committed:
Create .gitattributes

Add .module and .php.inc as php language

yorkshire-pudding committed:
Initial changes for port

Changed @package in header to refer to backdrop CMS
Removed help function
Removed definition of constants for variables.  The Backdrop method of getting module settings is different.
Changed drupal functions to backdrop equivalents.
Changed method of getting and setting variable
Change page_build to preprocess_page

yorkshire-pudding committed:
Update tawk_to.module

Fix small issue with line 111

alvinjohnsonso commented on tawk_to.module:
```diff
@@ -125,11 +108,11 @@ function tawk_to_page_build(&$page)
     }
 
-    $page['page_bottom']['tawkto_widget'] = array(
+	$variables['page_bottom'] .= array(
```
From what I saw in the example from [backdrop admin bar](https://github.com/backdrop/backdrop/commit/6490d6bfd7bf813e71266d128a66cbaf71284a1a#diff-76a37b8cdf08471bdcb862b66c903d2975eda8cf54c52627c21f2dbd06ed83f5R159), I think we can just simply say

```
$variables['page_bottom'] .= tawk_to_render_widget($page_id, $widget_id, $markupOptions);

yorkshire-pudding replied:
@alvinjohnsonso 
I've done that.  I can see the script at the bottom when I inspect elements:
```
<script type="text/javascript">
    var $_Tawk_API={},$_Tawk_LoadStart=new Date();
    (function(){
    var s1=document.createElement("script"),s0=document.getElementsByTagName("script")[0];
    s1.async=true;
    s1.src="https://embed.tawk.to/60ba37d1de99a4282a1b68ab/1f7bmk35i";
    s1.charset="UTF-8";
    s1.setAttribute("crossorigin","*");
    s0.parentNode.insertBefore(s1,s0);
    })();
    $_Tawk_API.visitor = {
            name  : "username",
            email : "name@domain.com",
        };
    var curpage = window.location.href;
    var hash = window.location.hash;
    $_Tawk_API.onLoad = function(){
        if (curpage.indexOf('admin/') != -1 || hash.indexOf('overlay')!= -1) {
            Tawk_API.hideWidget();
        }
    };
    </script>
```
but it doesn't seem to load on the page.

alvinjohnsonso replied:
Were there any network errors from tawk.to?

yorkshire-pudding replied:
Yes - there were:
register and https://va.tawk.to/v1/widget-settings?propertyId=60ba37d1de99a4282a1b68ab&widgetId=1f7bmk35i&sv=undefined

However, I tried on incognito and it worked, so guess it doesn't show for the person who set it up?

alvinjohnsonso replied:
That shouldn't be the case. Could you perhaps post the network request and response for `register`?

alvinjohnsonso replied:
Ahh but those error aren't caused by the plugin. It's from the widget itself.

alvinjohnsonso replied:
If we're able to save and remove the widget from the plugin's dashboard and the embed script tag shows on the front page, then that's basically it.

alvinjohnsonso replied:
Also, if the visibility settings are working properly

yorkshire-pudding replied:
It was the malwarebytes browser plugin - it is disabled in incognito but active otheriwse.  I trusted va.tawk.to and it was able to load.  Something to be aware of.  I don't know if tawk can get a global truest approval from malwarebytes.

yorkshire-pudding replied:
@alvinjohnsonso do you have any thoughts on what might need to be done to the menus? 
I'll have a look at the admin page

alvinjohnsonso replied:
Ahh as for the thing I said earlier from the call, I saw somewhere that the URLs instead of having full paths (ex. `some.hostname.com/admin/config/urls/settings`) they had query parameters to access the pages (ex. some.hostname.com?q=admin/config/urls/settings). Which I thought that we might need to change the ajax calls from the `admin.php.inc`.

Then I checked it out and seems like it's a setting that can be enabled/disabled from the backdrop admin page.
![image](https://user-images.githubusercontent.com/12890893/121693323-9cba7700-cafb-11eb-92b5-39c6a908d02c.png)

yorkshire-pudding replied:
Tested all functions and working.

alvinjohnsonso commented on tawk_to.module:
```diff
@@ -244,33 +227,35 @@ function tawk_to_set_widget()
+	$config = config('tawk_to.settings');
+	$config->clear('page_id');
+	$config->clear('widget_id');
+	$config->save;
```
Just some nitpicks, we're using 4 spaces for indentation instead of tabs 😁

yorkshire-pudding replied:
Good point.  I've applied those and committed the other changes

yorkshire-pudding committed:
Applied suggested changes

Modified variables['page_bottom'] and spacing rather than tabs

yorkshire-pudding committed:
Setting page styling and jquery hiding

Modified the html structure to match the html on comparable settings pages so the style will be consistent.
Added descriptions and changed labels to match backdrop terminology and make clearer what the different options mean.
Hidden the disabled controls to make the form simpler for users.

alvinjohnsonso commented on tawk_to.module:
```diff
@@ -2,47 +2,30 @@
-define('TAWKTO_WIDGET_PAGE_ID', 'tawkto_widget_page_id');
-define('TAWKTO_WIDGET_WIDGET_ID', 'tawkto_widget_widget_id');
-define('TAWKTO_WIDGET_OPTIONS', 'tawkto_widget_options');
-define('TAWKTO_WIDGET_USER_ID', 'tawkto_widget_user_id');
```
I do think we retain these constant variables and just change the value of them.
Ex.
`define('TAWKTO_WIDGET_PAGE_ID', 'tawkto_widget_page_id');` becomes `define('TAWKTO_WIDGET_PAGE_ID', 'page_id');`

This is so we can avoid repeating the hardcoded values below.

yorkshire-pudding replied:
@alvinjohnsonso I'm not quite sure of the benefits of doing this.  The constants will be longer than the strings they're replacing.  Please can you explain why this is better practice?

alvinjohnsonso replied:
Since the strings are repeating hard-coded values. I think it would be better to have those in a constant variable (even if it's a longer constant variable name). It'll be cleaner and if ever there's a time that we'll need to update the value of those, they're all just be in one place.

yorkshire-pudding replied:
Thanks for the explanation.  Done

alvinjohnsonso commented on tawk_to.module:
```diff
@@ -60,7 +43,7 @@ function tawk_to_page_build(&$page)
 
     // get visibility options
-    $options = variable_get(TAWKTO_WIDGET_OPTIONS, '');
+    $options = config_get('tawk_to.settings','widget_options');
```
We may also want to add the config name to the constants declaration. Something like `define('TAWKTO_CONFIG_NAME', 'tawk_to.settings');` so we can just define it once.

yorkshire-pudding replied:
@alvinjohnsonso I'm not quite sure of the benefits of doing this.  The constants will be longer than the strings they're replacing.  Please can you explain why this is better practice?

yorkshire-pudding replied:
Thanks for the explanation. Done

alvinjohnsonso commented on views/admin.php.inc:
```diff
@@ -1,183 +1,173 @@
-                                Put each URL in a new line.
-                                </p>
+{   ob_start();
+    ?><script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.0/jquery.min.js"></script>
```
No need to add jquery cdn since backdrop cms loads up jquery by default

yorkshire-pudding replied:
Removed.  I had removed it, but ending up adding back in when trying to troubleshoot an issue and forgot to remove it.

alvinjohnsonso commented on views/admin.php.inc:
```diff
@@ -1,183 +1,173 @@
+                                                    foreach ($whitelist as $page) { echo $page."\r\n"; }
+                                            } else {
+                                            } ?></textarea>
+                                            <div class="grippie"></div>
```
The grippie doesn't work so it can be removed if not needed. It might confuse people that try to drag it down.

yorkshire-pudding replied:
Removed.  I saw it work on another admin page and thought it was a good idea but the bit of JS powering it does not seem to be universal

alvinjohnsonso commented on views/admin.php.inc:
```diff
@@ -1,183 +1,173 @@
+                                        <?php
+                                            $checked = false;
+                                            if (!is_null($display_opts)) {
+                                                if (!$display_opts->show_onfrontpage) {
```
This should be `if ($display_opts->show_onfrontpage) {`

yorkshire-pudding replied:
Done

alvinjohnsonso commented on views/admin.php.inc:
```diff
@@ -1,183 +1,173 @@
+                                        <?php
+                                            $checked = false;
+                                            if (!is_null($display_opts)) {
+                                                if (!$display_opts->show_oncategory) {
```
This should be `if ($display_opts->show_oncategory) {`

alvinjohnsonso replied:
Or if we're changing to `taxonomy` (which makes much more sense since Backdrop uses `Taxonomy`), we can change `show_oncategory` to `show_ontaxonomy`.

yorkshire-pudding replied:
Done. Good suggestion

alvinjohnsonso commented on views/admin.php.inc:
```diff
@@ -1,183 +1,173 @@
+                                                foreach ($whitelist as $page) { echo $page."\r\n"; }
+                                            } else {
+                                            } ?></textarea>
+                                            <div class="grippie"></div>
```
The grippie doesn't work so it can be removed if not needed. It might confuse people that try to drag it down.

yorkshire-pudding replied:
Removed.  I saw it work on another admin page and thought it was a good idea but the bit of JS powering it does not seem to be universal

alvinjohnsonso commented on views/admin.php.inc:
```diff
@@ -227,17 +217,23 @@ function tawk_to_render_widget_iframe($base_url, $iframe_url, $widget=array(), $
         jQuery(document).ready(function() {
             if(jQuery("#always_display").prop("checked")){
                 jQuery('.show_specific').prop('disabled', true);
+                $(".div_show_specific").hide();
```
Use `jQuery` instead of `$`

yorkshire-pudding replied:
Good tip.  When I recently learned about jQuery all the guides used the dollar.  When I searched for jQuery or dollar I found reference to the possibility for conflict where another framework (i.e. PHP in this case) uses dollar for variables.  I've made this change.

alvinjohnsonso commented on views/admin.php.inc:
```diff
@@ -1,183 +1,173 @@
+                                <?php if (!$sameUser) { ?>
+                                    <div id="widget_already_set" class="alert alert-warning">Notice: Widget already set by other user</div>
+                                <?php } ?>
+                                <iframe id="tawkIframe" src="" style="min-height: 275px; width: 80%; border: none; margin: 5px 0; padding: 10px; background: #FFF;"></iframe>
```
`padding` and `background` can be removed. Then, `width` can be set to `100%`.

yorkshire-pudding replied:
Done

alvinjohnsonso commented on views/admin.php.inc:
```diff
@@ -1,183 +1,173 @@
+                                            <textarea class="form-textarea hide_specific" name="hide_oncustom" id="hide_oncustom" cols="40" rows="6"><?php if (!empty($display_opts->hide_oncustom)) {
+                                                    $whitelist = json_decode($display_opts->hide_oncustom);                                                     
+                                                    foreach ($whitelist as $page) { echo $page."\r\n"; }
+                                            } else {
```
I think we can remove the `else` block here.

yorkshire-pudding replied:
Done

yorkshire-pudding committed:
Several changes suggested by Alvin

Removal of surplus else{} blocks
Change to iframe styling
jQuery identifier from dollar to jQuery
Remove non-functioning grippie divs
Change from show_oncategory to show_ontaxonomy
Condition corrections
Removal of jquery cdn load

yorkshire-pudding committed:
Constants for config settings and some comment changes

Replaced values in config_get and config_set with constants.
Updated and improved some comments for clarity and to reflect minor changes with Backdrop.
Removed some lines of code that were commented out

yorkshire-pudding committed:
Exceptions just use page URL not domain

Change to the config to allow users to just put the page url from the Backdrop root without the base_url (i.e. the domain) (and the parent folder root is not at domain root).  Made clear in the description.  This will simplify for users, but also avoid problems with people moving config from dev/staging/test to production.

yorkshire-pudding committed:
Fix issue with remove_widget function

After testing removing widget made correction to fix issue.

yorkshire-pudding committed:
Removed styling from admin to CSS file

Removed styling from admin.inc.php
Created css folder and added main.css with styling in
Added classes for referring to elements styled by the CSS
Added reference in module file to load the CSS

yorkshire-pudding committed:
Change success message from ID to class

Identify change by class rather than ID

yorkshire-pudding committed:
Styling changes and CSS name change

Whitespace and LF standards applied
Changed name of CSS file

yorkshire-pudding committed:
Moved admin and widget to parent

Moved admin and widget to parent.

yorkshire-pudding commented:
@alvinjohnsonso I think these are all the changes needed.  Moving the JS out didn't work because of the PHP in there.  I've tidied up the files a bit to match the recommended file naming structure.

alvinjohnsonso commented on tawk_to.admin.inc:
```diff
@@ -0,0 +1,260 @@
+    </div>
+    <script type="text/javascript">
+        var currentHost = window.location.protocol + "//" + window.location.host;
+        var url = "<?php echo $iframe_url; ?>&pltf=drupal&pltfv=7&parentDomain=" + currentHost;
```
With regards to this issue https://github.com/tawk/tawk-backdrop/issues/2, you can go ahead and change the `pltf=drupal` to `pltf=backdrop` and remove `pltfv=7`

yorkshire-pudding replied:
Done

alvinjohnsonso commented:
> @alvinjohnsonso I think these are all the changes needed. Moving the JS out didn't work because of the PHP in there. I've tidied up the files a bit to match the recommended file naming structure.

@yorkshire-pudding what's the reason for moving the `tawk_to.admin.inc` and `tawk_to.widget.inc` out of the `views` folder?

yorkshire-pudding commented:
@alvinjohnsonso looking at other backdrop modules admin and other included files (except classes, CSS and js) are mostly in the parent folder. I also think have a folder called views in the context of backdrop which has views as a structural config item (like Drupal) could be misleading at first glance and suggest that the module includes Views

alvinjohnsonso commented:
Ahhh I see. Alright, I think we can keep it there since I just saw this in the Backdrop docs

https://docs.backdropcms.org/creating-modules

yorkshire-pudding committed:
Update platform in iframe URL

Update platform from Drupal version 7 to Backdrop (no version)

alvinjohnsonso commented:
@yorkshire-pudding took a second look at the docs and saw that we can move the `tawk_to.admin.inc` and `tawk_to.widget.inc` inside an `includes` folder since they said that

`If there are multiple of these files however, it is reasonable to move these files to an includes folder in the module root`.

By doing this, we can separate them from the root module files and have a much more organized structure.

yorkshire-pudding commented:
> @yorkshire-pudding took a second look at the docs and saw that we can move the `tawk_to.admin.inc` and `tawk_to.widget.inc` inside an `includes` folder since they said that
> 
> `If there are multiple of these files however, it is reasonable to move these files to an includes folder in the module root`.
> 
> By doing this, we can separate them from the root module files and have a much more organized structure.

@alvinjohnsonso but the previous sentence says ".theme.inc, .admin.inc, and.pages.inc files are recommended to be kept in the root of the module folder to allow ease of recognition of the way the module likely works. This makes it more obvious that the module provides theme functions, has an admin page, and probably provides display pages in some way. "

We only have one admin inc and one widget inc so I think it is reasonable to keep in root. Many other modules have similar and only class includes are put in an includes folder.

yorkshire-pudding committed:
Added cache_flush to set_widget and remove_widget

Added `cache_flush('page');` function call to set_widget and remove_widget functions as an alternative solution to the issue described in https://github.com/backdrop-ops/contrib/issues/534#issuecomment-865092111

yorkshire-pudding commented:
@alvinjohnsonso I have added the `cache_flush('page');` function call to set_widget and remove_widget functions as an alternative solution to the issue described in https://github.com/backdrop-ops/contrib/issues/534#issuecomment-865092111

alvinjohnsonso commented:
@yorkshire-pudding tested it and seems to be not working as intended. I think we can just wait for this PR to be merged https://github.com/backdrop/backdrop/pull/3475 and use it once it's available

yorkshire-pudding commented:
> @yorkshire-pudding tested it and seems to be not working as intended. I think we can just wait for this PR to be merged https://github.com/backdrop/backdrop/pull/3475 and use it once it's available

@alvinjohnsonso that's strange; I tested it several times, comparing the outcome with the lines enabled and disabled and the difference was clear. What steps did you carry out and what was the result you got compared to what you were expecting?

I presume you were expecting the widget to appear or disappear on next page reload for anonymous users.

alvinjohnsonso commented:
@yorkshire-pudding All I did was when I selected a widget, I opened up a new browser and went to the homepage of the website. The widget loads up. Then I changed it to another widget, refreshed the page and the widget didn't change. The same thing goes when I removed the widget.

yorkshire-pudding commented:
> @yorkshire-pudding All I did was when I selected a widget, I opened up a new browser and went to the homepage of the website. The widget loads up. Then I changed it to another widget, refreshed the page and the widget didn't change. The same thing goes when I removed the widget.

@alvinjohnsonso this is very odd. I've double checked that I have the same code on my test site. I've tried in Firefox, edge, mobile Chrome, mobile duck duck go and mobile browser from VMware on my employer phone. Widget appears, disappears and changes on page refresh. I presume there is nothing like cloud flare in front of your test site?

alvinjohnsonso commented:
> > @yorkshire-pudding All I did was when I selected a widget, I opened up a new browser and went to the homepage of the website. The widget loads up. Then I changed it to another widget, refreshed the page and the widget didn't change. The same thing goes when I removed the widget.
> 
> @alvinjohnsonso this is very odd. I've double checked that I have the same code on my test site. I've tried in Firefox, edge, mobile Chrome, mobile duck duck go and mobile browser from VMware on my employer phone. Widget appears, disappears and changes on page refresh. I presume there is nothing like cloud flare in front of your test site?

@yorkshire-pudding yup. I'm actually testing it in a local environment. Here's a loom video of me testing it just now.
https://www.loom.com/share/d3c8c5e6f48c4e81879bf79c2e2f0c24

alvinjohnsonso commented:
> > > @yorkshire-pudding All I did was when I selected a widget, I opened up a new browser and went to the homepage of the website. The widget loads up. Then I changed it to another widget, refreshed the page and the widget didn't change. The same thing goes when I removed the widget.
> > 
> > 
> > @alvinjohnsonso this is very odd. I've double checked that I have the same code on my test site. I've tried in Firefox, edge, mobile Chrome, mobile duck duck go and mobile browser from VMware on my employer phone. Widget appears, disappears and changes on page refresh. I presume there is nothing like cloud flare in front of your test site?
> 
> @yorkshire-pudding yup. I'm actually testing it in a local environment. Here's a loom video of me testing it just now.
> https://www.loom.com/share/d3c8c5e6f48c4e81879bf79c2e2f0c24
> 
> Could it be that the `cached pages for anonymous users` are turned off in your performance settings?

@yorkshire-pudding terribly sorry. Just did a fresh install of the backdrop CMS and now the cache flushing works perfectly. Not sure what happened to the earlier copy but it works fine now.

yorkshire-pudding commented:
@alvinjohnsonso that's a relief; I was going to spin up a fresh install to double check.

yorkshire-pudding committed:
Correct comment in remove_widget

Corrected comment so refers to widget disappearing at next page load rather than appearing

alvinjohnsonso commented on tawk_to.module:
```diff
@@ -310,7 +306,7 @@ function tawkto_widget_options()
         }
     }
-    variable_set(TAWKTO_WIDGET_OPTIONS, json_encode($jsonOpts));
+    config_set(TAWK_TO_CONFIG_NAME, TAWK_TO_WIDGET_OPTIONS, json_encode($jsonOpts));
```
@yorkshire-pudding can we also add the cache flushing here? 😁

yorkshire-pudding replied:
Done

yorkshire-pudding committed:
flush cache when update options

Added call to cache_flush('page') to widget_options so changes to settings for where to show or privacy will update at next page load.

yorkshire-pudding commented:
@alvinjohnsonso Are we able to merge this pull request and #3 ?  And then presumably we can apply to have it included.

yorkshire-pudding commented:
@alvinjohnsonso it would be good to get this completed.  Are you able to record review and merge?

alvinjohnsonso commented:
Hi @yorkshire-pudding, apologies for the late response. Will try to have this merged by next week and release it to the backdrop-contrib group.

yorkshire-pudding commented:
Hi @alvinjohnsonso, have you had chance to look at this yet?
